packages/financial-templates-lib/src/price-feed/TraderMadePriceFeed.js

```diff
@@ -0,0 +1,255 @@
+   * @param {Object} networker Used to send the API requests.
+   * @param {Function} getTime Returns the current time.
+   * @param {Integer} minTimeBetweenUpdates Min number of seconds between updates. If update() is called again before
+   *      this number of seconds has passed, it will be a no-op.
+   * @param {Number} priceFeedDecimals Number of priceFeedDecimals to use to convert price to wei.
+
+   */
+  constructor(logger, web3, apiKey, pair, lookback, networker, getTime, minTimeBetweenUpdates, priceFeedDecimals = 6) {
+    super();
+    this.logger = logger;
+    this.web3 = web3;
+
+    this.apiKey = apiKey;
+    this.pair = pair;
+
+    this.lookback = lookback;
+    this.networker = networker;
+    this.getTime = getTime;
+    this.minTimeBetweenUpdates = minTimeBetweenUpdates;
+
+    this.toBN = this.web3.utils.toBN;
+
+    this.priceFeedDecimals = priceFeedDecimals;
+    this.convertPriceFeedDecimals = number => {
+      // Converts price result to wei
+      // returns price conversion to correct decimals as a big number
+      return this.toBN(parseFixed(number.toString(), priceFeedDecimals).toString());
+    };
+  }
+
+  getCurrentPrice() {
+    return this.currentPrice;
+  }
+
+  getHistoricalPrice(time, verbose = false) {
+    if (this.lastUpdateTime === undefined) {
+      return null;
+    }
+
+    // Set first price time in `historicalPrices` to first non-null price.
+    let firstPriceTime;
+    for (let p in this.historicalPrices) {
+      if (this.historicalPrices[p] && this.historicalPrices[p].openTime) {
+        firstPriceTime = this.historicalPrices[p];
+        break;
+      }
+    }
+
+    // If there are no valid price time, return null.
+    if (!firstPriceTime) {
+      return null;
+    }
+
+    // If the time is before the first piece of data in the set, return null because
+    // the price is before the lookback window.
+    if (time < firstPriceTime.openTime) {
+      return null;
+    }
+
+    // historicalPrices are ordered from oldest to newest.
+    // This finds the first priceTime whose closeTime is after the provided time.
+    const match = this.historicalPrices.find(price => {
+      return time < price.closeTime;
+    });
+
+    // If there is no match, that means that the time was past the last data point.
+    // In this case, the best match for this price is the current price.
+    if (match === undefined) {
+      let returnPrice = this.currentPrice;
+      if (verbose) {
+        console.group(`\n(${this.pair}) No OHLC available @ ${time}`);
+        console.log(
+          `- ✅ Time is later than earliest historical time, fetching current price: ${this.web3.utils.fromWei(
+            returnPrice.toString()
+          )}`
+        );
+        console.log(
+          `- ⚠️  If you want to manually verify the specific exchange prices, you can make a GET request to: \n- https://marketdata.tradermade.com/api/v1/live?currency=${this.pair}&api_key={api-key}`
```

Comment:
Does this work without an `apiKey`? If not can you add it to the list of required params in `CreatePriceFeed`

Reply:
sure,